### DISCUSSION TITLE
poor man's recurrence

### DISCUSSION THREAD

**Discussion by gour** [GothenburgBitFactory/taskwarrior#2846] [Community support] [answered]
Hello,

lack of proper recurrence support is my main pain in using Taskwarrior, but despite of that it's the most productive tool for task management. Considering I've a lot of recurring tasks, I wonder what would be the pro/cons of converting them to non-recurring ones? For example, **instead** of completing one when it's due or when I manage to finish it, to just **modify** its due date with e.g., `+1w` for a weekly task?

It is not perfect, but, maybe, better than  Taskwarrior creating new instances with incorrect due dates?

**Comment by Arwalk** [comment]
This can probably be done with a custom uda that would hold your recurring value (ie `+1w`) and some on-modify hooks. Need some help to do that?

**Reply by gour** [reply]
> This can probably be done with a custom uda that would hold your recurring value (ie `+1w`) and some on-modify hooks. Need some help to do that?

I believe I did try with on-modify hook in the past, but without uda, so if you believe there is some workaround to accomplish it, I'm all ears...

**Reply by Arwalk** [reply, marked as answer]
Ok i think i have something that should work along the lines of what you're looking for (or at least easily adaptable to what you want).

First, i added a uda in my `.taskrc` 

![image](https://user-images.githubusercontent.com/4776412/173233169-4bffcb98-d6db-48a6-90cc-8ee78289ef15.png)

I wrote this simple `on-modify.crecurr.py` script.

this `crecurr` for `custom recurrence` is supposed to have a string in the format of {value}{unit}

```python
#!/usr/bin/env python3

import sys
import json
import datetime
import re

DATETIME_FORMAT = "%Y%m%dT%H%M%SZ"

original_task = json.loads(sys.stdin.readline())
modified_task = json.loads(sys.stdin.readline())

crecurr = modified_task.get("crecurr", None)

# if we are actively making a task with crecurr "done"
if crecurr and modified_task["status"] == "completed":
    # reset it to pending state
    modified_task["status"] = "pending"

    # extract the current due date
    due = modified_task["due"]
    due = datetime.datetime.strptime(due, DATETIME_FORMAT)

    # extract the different values in crecurr as a list of (value, unit)
    crecurr = crecurr.split()
    crecurr = [[y for y in re.split('(\d+)', x) if y] for x in crecurr]

    delta = {}

    # build ou delta using the values in crecurr
    for value, unit in crecurr:
        if unit == 'w':
            delta["weeks"] = int(value)
        if unit == 'd':
            delta["days"] = int(value)
        # etc etc fill with all the cases you want to manage

    # expand it to an actual timedelta object
    delta = datetime.timedelta(**delta)

    # update the current due date
    due = due + delta
    modified_task["due"] = due.strftime(DATETIME_FORMAT)

print(json.dumps(modified_task))

sys.exit(0)
```

Now for an example, here is a task with a due date and our expected recurrence

![image](https://user-images.githubusercontent.com/4776412/173233546-b270d628-876e-479e-ae03-93573f586864.png)

Now completed the usual way

![image](https://user-images.githubusercontent.com/4776412/173233577-62773190-d2d9-4737-94a6-9cb4acbe2ba3.png)

But if i inspect it again...

![image](https://user-images.githubusercontent.com/4776412/173233619-d6476449-1123-4411-b165-07c0b1ae4094.png)

It only updated the due date.

**Reply by gour** [reply]
> Ok i think i have something that should work along the lines of what you're looking for (or at least easily adaptable to what you want).

Thank you very much!

Maybe I do miss something, but it does not work as expected...

> First, i added a uda in my `.taskrc`

```
$ task udas

Name     Type   Label             Allowed Values Default Usage Count
crecurr  string Custom Recurrence                        0          
priority string Priority          H,M,L,                 349
reviewed date   Reviewed                                 23         

3 UDAs defined
```

> I wrote this simple `on-modify.crecurr.py` script.

```
$  ll task/hooks
total 4,0K
-rwxr-xr-x 1 gour gour 1,3K lip  14 11:45 on-modify.crecurr.py*

```

> this `crecurr` for `custom recurrence` is supposed to have a string in the format of {value}{unit}

```
$ task long 180

ID  Created     Mod  Recur Due         Description
180 14.6.2022.  1h     P7D 13.6.2022.  testrecur  

```
> Now for an example, here is a task with a due date and our expected recurrence

My task does not have `Custom Recurrence`

```
task info 180

Name            Value                                                                            
ID              180                                                                              
Description     testrecur                                                                        
Status          Pending                                                                          
Recurrence      1w
Parent task     762c6bd9-8d45-4f58-8618-df035ca9ecf0                                             
Mask Index      0
Recurrence type periodic                                                                         
Entered         2022-06-14 11:46:18 (1h)
Due             2022-06-13 00:00:00                                                              
Last modified   2022-06-14 11:46:08 (1h)
Virtual tags    CHILD INSTANCE MONTH OVERDUE PENDING QUARTER READY UNBLOCKED WEEK YEAR YESTERDAY 
UUID            035444e0-19b9-4d5a-9450-d28b7effbb27
Urgency         9.504                                                                            

    due  0.792 *   12 =    9.5
                        ------
                         9.504
```
> Now completed the usual way

```
$ task 180 done
Completed task 035444e0 'testrecur'.
Completed 1 task.
You have more urgent tasks.
 $  task long test

ID  Created     Mod  Recur Due         Description
180 14.6.2022.  1h     P7D 20.6.2022.  testrecur  

1 task
There are 2 local changes.  Sync required.
$ task info 180

Name            Value                                                                
ID              180                                                                  
Description     testrecur                                                            
Status          Pending                                                              
Recurrence      1w
Parent task     762c6bd9-8d45-4f58-8618-df035ca9ecf0                                 
Mask Index      1
Recurrence type periodic                                                             
Entered         2022-06-14 11:46:18 (1h)
Due             2022-06-20 00:00:00                                                  
Last modified   2022-06-14 11:46:08 (1h)
Virtual tags    CHILD DUE INSTANCE LATEST MONTH PENDING QUARTER READY UNBLOCKED YEAR 
UUID            c77461c1-d1c4-42c3-a180-192ec1beb28f
Urgency         6.304                                                                

    due  0.525 *   12 =    6.3
                        ------
                         6.304
```

What do I miss?

**Reply by Arwalk** [reply]
You need to add the attribute "crecurr" to your task, just like you used to add the "recur" attribute before.

Basically, instead of `task add recur:1w due:tomorro "testtask"`, you do `task add crecurr:1w due:tomorrow "testtask"` 

I wouldn't mix the normal recur and this customized one, so don't hesitate to remove all the original recurring tasks and replace them.

**Reply by gour** [reply]
> You need to add the attribute "crecurr" to your task, just like you used to add the "recur" attribute before.

Ahh, now it makes sense. Thank you. :wink: 

> Basically, instead of `task add recur:1w due:tomorro "testtask"`, you do `task add crecurr:1w due:tomorrow "testtask"`

OK, I did it and here is the sequence:

```
tasksh> add testrecur due:yesterday crecurr:1w
[task add testrecur due:yesterday crecurr:1w]
Created task 179.
tasksh> 179 info
[task 179 info]

Name              Value                                                                        
ID                179                                                                          
Description       testrecur                                                                    
Status            Pending                                                                      
Entered           2022-06-14 15:14:31 (9s)
Due               2022-06-13 00:00:00                                                          
Last modified     2022-06-14 15:14:31 (9s)
Virtual tags      LATEST MONTH OVERDUE PENDING QUARTER READY UDA UNBLOCKED WEEK YEAR YESTERDAY 
UUID              f28ea60f-a8a2-40f2-8faf-cd907cfd6b63
Urgency           9.548                                                                        
Custom Recurrence 1w
```

Everything is as expected...

Now, let's complete that task:

```
tasksh> 179 done
[task 179 done]
Completed task f28ea60f 'testrecur'.
Completed 1 task.
You have more urgent tasks.
```

and inspect new instance:

```
tasksh> 179 info
[task 179 info]

Name              Value                                                     
ID                179                                                       
Description       testrecur                                                 
Status            Pending                                                   
Entered           2022-06-14 15:14:31 (27s)
Due               2022-06-20 00:00:00                                       
Last modified     2022-06-14 15:14:49 (9s)
Virtual tags      DUE LATEST MONTH PENDING QUARTER READY UDA UNBLOCKED YEAR 
UUID              f28ea60f-a8a2-40f2-8faf-cd907cfd6b63
Urgency           6.348                                                     
Custom Recurrence 1w
```

So, I'd expect that the new instance would be due on `2022-06-21 00:00:00` and not on `20th`?

Any hint?

> I wouldn't mix the normal recur and this customized one, so don't hesitate to remove all the original recurring tasks and replace them.

Sure. :-)

**Reply by Arwalk** [reply]
Hmm, for me the behavior is ok, due date was `2022-06-13 00:00:00`, if you add a week it makes `2022-06-20 00:00:00`. My script updates the original due date. If you want to have it 1 week after the time you completed the task, the block after the `if` in the script could look like this

```python
    # reset it to pending state
    modified_task["status"] = "pending"


    # extract the current date
   # /!\ NEXT LINE CHANGED
    due = datetime.now().replace(hour=0, minute=0, second=0)

    # extract the different values in crecurr as a list of (value, unit)
    crecurr = crecurr.split()
    crecurr = [[y for y in re.split('(\d+)', x) if y] for x in crecurr]

    delta = {}

    # build ou delta using the values in crecurr
    for value, unit in crecurr:
        if unit == 'w':
            delta["weeks"] = int(value)
        if unit == 'd':
            delta["days"] = int(value)
        # etc etc fill with all the cases you want to manage

    # expand it to an actual timedelta object
    delta = datetime.timedelta(**delta)

    # update the current due date
    due = due + delta
    modified_task["due"] = due.strftime(DATETIME_FORMAT)
```


Overall, you can probably hack into the python script at this point, and make it behave as you wish!

**Reply by gour** [reply]
> If you want to have it 1 week after the time you completed the task, the block after the `if` in the script could look like this

Yes, I want that the new instance have the same period, e.g. "mow the meadow" once in a week, so if I'm a bit late by completing the task two days later, I want new instance in seven days and not five. :-)

>     due = datetime.now().replace(hour=0, minute=0, second=0)

`due = datetime.datetime.now().replace(hour=0, minute=0, second=0)` was the only required change to make it work!

> Overall, you can probably hack into the python script at this point, and make it behave as you wish!

Thanks a lot for your help, it looks that it can solve long problem I had with Taskwarrior due to lack of proper "recurrence overhaul". :-)
